third_party/move/move-vm/runtime/src/data_cache.rs

```diff
@@ -27,19 +27,43 @@ use move_vm_types::{
 };
 use std::collections::btree_map::{BTreeMap, Entry};
 
+enum CachedInformation {
+    Value(GlobalValue),
+    SizeOnly(Option<u64>),
+}
+
+impl CachedInformation {
+    fn value_mut(&mut self) -> PartialVMResult<&mut GlobalValue> {
+        match self {
+            CachedInformation::Value(v) => Ok(v),
+            CachedInformation::SizeOnly(_) => Err(PartialVMError::new_invariant_violation("Data is not cached"))
+        }
+    }
+}
+
 /// An entry in the data cache, containing resource's [GlobalValue] as well as additional cached
 /// information such as tag, layout, and a flag whether there are any delayed fields inside the
 /// resource.
 pub(crate) struct DataCacheEntry {
     struct_tag: StructTag,
     layout: MoveTypeLayout,
     contains_delayed_fields: bool,
-    value: GlobalValue,
+    value: CachedInformation,
 }
 
 impl DataCacheEntry {
-    pub(crate) fn value(&self) -> &GlobalValue {
-        &self.value
+    pub(crate) fn maybe_value(&self) -> Option<&GlobalValue> {
+        match &self.value {
+            CachedInformation::SizeOnly(_) => None,
+            CachedInformation::Value(v) => Some(v),
+        }
+    }
+
+    pub(crate) fn exists(&self) -> PartialVMResult<bool> {
+        match &self.value {
+            CachedInformation::SizeOnly(e) => Ok(e.unwrap_or(0) != 0),
```

Comment:
With the fix to `executor_view_with_change_set`, this should work with just `e.is_some()` I think